### PR TITLE
Add Error Prone check: ReturnValueIgnored

### PR DESCRIPTION
**Pull request by ksobolew**
The individual instances of where the check reports errors are mostly suppressed, as there's a couple of places where we rely on side effects, instead of the return value. In one of these there is a parameter passed as a `Function`, where a `Consumer` would be a better interface to use, as the result of the function is ignored. Another interesting case is where there is a `Function` provided as a customizer in the builder pattern, but the ways it's invoked it only relies on side effects (modifying the builder in-place) and the result is ignored; in this case it's prudent to actually use the result as the builder object to continue with, even if in practice it's exactly the same as the input object.

### CODE REVIEW

File: testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java [hunk 1]
```diff
@@ -49,13 +48,15 @@ public void checkPlanIsDeterministic(String sql)
 
     public void checkPlanIsDeterministic(Session session, String sql)
     {
-        IntStream.range(1, MINIMUM_SUBSEQUENT_SAME_PLANS)
-                .mapToObj(attempt -> getPlanText(session, sql))
-                .map(planEquivalenceFunction)
-                .reduce((previous, current) -> {
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java:55]: You can keep it the stream way, just replace `reduce` with `forEach`

**Review by findepi — APPROVED** [review]
%  https://github.com/trinodb/trino/pull/6898/files#r576268294

File: testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java (hunk 1 above)
Comment (ksobolew) [reply testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java:55]: Well, this was coded using `reduce` because it depends on access to adjacent elements in the stream (to make sure that all elements are equal). With `forEach` the handler would have to keep state between invocations. @kokosing suggested to turn it into a loop, and I too think that when we rely on side effects the loop way is probably a better idea anyway.

Comment (findepi) [reply testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java:55]: I missed that part. You're right.

**Comment by findepi** [comment]
please rebase

File: core/trino-main/src/main/java/io/trino/execution/SqlTask.java [hunk 2]
```diff
@@ -90,14 +90,14 @@
     private final AtomicReference<TaskHolder> taskHolderReference = new AtomicReference<>(new TaskHolder());
     private final AtomicBoolean needsPlan = new AtomicBoolean(true);
 
-    public static SqlTask createSqlTask(
+    static SqlTask createSqlTask(
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/SqlTask.java:93]: unrelated?

File: testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java [hunk 3]
```diff
@@ -49,13 +48,15 @@ public void checkPlanIsDeterministic(String sql)
...
-                    assertEquals(previous, current);
-                    return current;
-                });
+        String previous = null;
+        for (int attempt = 1; attempt < MINIMUM_SUBSEQUENT_SAME_PLANS; attempt++) {
+            String planText = getPlanText(session, sql);
+            String current = planEquivalenceFunction.apply(planText);
+            if (previous != null) {
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java:55]: Base it on `attempt` value, so you don't ignore it when `previous` is null on not the first invocation.

Or perhaps init the previous before the loop body (and rename):
```
String first = planEquivalenceFunction.apply(getPlanText(session, sql));
```

File: testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java [hunk 4]
```diff
@@ -49,13 +48,15 @@ public void checkPlanIsDeterministic(String sql)
...
     {
-        IntStream.range(1, MINIMUM_SUBSEQUENT_SAME_PLANS)
-                .mapToObj(attempt -> getPlanText(session, sql))
-                .map(planEquivalenceFunction)
-                .reduce((previous, current) -> {
-                    assertEquals(previous, current);
-                    return current;
-                });
+        String previous = null;
+        for (int attempt = 1; attempt < MINIMUM_SUBSEQUENT_SAME_PLANS; attempt++) {
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java:52]: We're checking that MINIMUM_SUBSEQUENT_SAME_PLANS-1 plans are the same, aren't we?

this is pre-existing, but since you're rewriting it...

**Review by findepi — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/SqlTask.java (hunk 2 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/execution/SqlTask.java:93]: It's one of the methods that gets the API changed, I just wanted to hide it somewhat to make sure that the impact is limited. Should I leave it `public`?

File: testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java (hunk 4 above)
Comment (ksobolew) [reply testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java:52]: Yes, this is something that bothered me, too. Looks like an off-by-one to me, so I wanted a second opinion of someone who can judge the intent more precisely (see https://github.com/trinodb/trino/pull/6879/files#r576112688). I'll "fix" it and see what people think (though a regression test would be nice, too)

File: core/trino-main/src/main/java/io/trino/execution/SqlTask.java (hunk 2 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/SqlTask.java:93]: The public or not is orthogonal to this PR, so keep it as it was.

BTW trino-main has no "public API", it's not a library

File: testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java (hunk 4 above)
Comment (findepi) [reply testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java:52]: regression test is not applcable -- this is just assertion coide in a test; there is no way to test that (other than contrived test cases which wouldn't be useful)

Comment (ksobolew) [reply testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java:52]: Yeah, I figured as much. It's just making the existing tests a little stronger than before

File: testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java (hunk 3 above)
Comment (ksobolew) [reply testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java]: I took the latter approach, which also fixes the apparent off-by-one in the number of iterations.

File: core/trino-main/src/main/java/io/trino/execution/SqlTask.java (hunk 2 above)
Comment (ksobolew) [reply core/trino-main/src/main/java/io/trino/execution/SqlTask.java]: Sure, I backed out this change.

File: testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java [hunk 5]
```diff
@@ -48,13 +48,10 @@ public void checkPlanIsDeterministic(String sql)
 
     public void checkPlanIsDeterministic(Session session, String sql)
     {
-        String previous = null;
+        String previous = planEquivalenceFunction.apply(getPlanText(session, sql));
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java:51]: remove `previous = current;` below

Comment (ksobolew) [reply testing/trino-testing/src/main/java/io/trino/testing/PlanDeterminismChecker.java:51]: You're right, since they are all supposed to be identical, there's no point in tracking the previous values. Removed.

**Comment by findepi** [comment]
thanks, please squash
